rust/rope/src/spans.rs

```diff
@@ -181,6 +181,108 @@ impl<T: Clone + Default> Spans<T> {
         builder.build()
     }
 
+    /// Creates a new Spans instance by merging spans from `other` with `self`,
+    /// using a closure to transform values.
+    ///
+    /// New spans are created from non-overlapping regions of existing spans,
+    /// and by combining overlapping regions into new spans. In all cases,
+    /// new values are generated by calling a closure that transforms the
+    /// value of the existing span or spans.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self` and `other` have different lengths.
+    ///
+    pub fn merge<F, O>(&self, other: &Self, mut f: F) -> Spans<O>
+        where F: FnMut(&T, Option<&T>) -> O,
+              O: Clone + Default,
+    {
+        //TODO: confirm that this is sensible behaviour
+        assert_eq!(self.len(), other.len());
+        let mut sb = SpansBuilder::new(self.len());
+
+        // red/blue is just a better name than one/two or me/other
+        let mut iter_red = self.iter();
+        let mut iter_blue = other.iter();
+
+        let mut next_red = iter_red.next();
+        let mut next_blue = iter_blue.next();
+
+        loop {
+            // exit conditions:
+            if next_red.is_none() && next_blue.is_none() {
+                // all merged.
+                break
+            } else if next_red.is_none() != next_blue.is_none() {
+                // one side is exhausted; append remaining items from other side.
+                let iter = if next_red.is_some() { iter_red } else { iter_blue };
+                // add this item
+                let (iv, val) = next_red.or(next_blue).unwrap();
+                sb.add_span(iv, f(val, None));
+
+                for (iv, val) in iter {
+                    sb.add_span(iv, f(val, None))
+                }
+                break
+            }
+
+            // body:
+            let (mut red_iv, red_val) = next_red.unwrap();
+            let (mut blue_iv, blue_val) = next_blue.unwrap();
+
+            if red_iv.intersect(blue_iv).is_empty() {
+                // the spans do not overlap. Add and advance the leftmost.
+                if red_iv.is_before(blue_iv.start()) {
+                    sb.add_span(red_iv, f(red_val, None));
+                    next_red = iter_red.next();
+                } else {
+                    sb.add_span(blue_iv, f(blue_val, None));
+                    next_blue = iter_blue.next();
+                }
+                continue
+            }
+            assert!(!red_iv.intersect(blue_iv).is_empty());
+
+            // if these two spans do not share a start point, create a new span from
+            // the prefix of the leading span.
+            if red_iv.start() < blue_iv.start() {
+                let iv = red_iv.prefix(blue_iv);
+                sb.add_span(iv, f(red_val, None));
+                red_iv = red_iv.suffix(iv);
+            }
+            else if blue_iv.start() < red_iv.start() {
```

Comment:
yea I think I originally broke this line because it looked visually dense without, but that doesn't seem super necessary in hindsight, and I can always use comments or something if I feel the need to break it up.